### ISSUE TITLE
[2019 ICLR] Approximating CNNs with Bag-of-local-Features models works surprisingly well on ImageNet

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#67]
Receptive field를 작게 가지는 bag-of-feature model인 BagNet을 제안한 논문

네트워크와 실험만 설명

## 1. Network Architecture
<img width="647" alt="image" src="https://user-images.githubusercontent.com/49643709/167383861-304bc6fe-e14e-45ba-ae91-a23c64bf5b8b.png">
<img width="636" alt="image" src="https://user-images.githubusercontent.com/49643709/167383806-0acfbc91-5b38-45e0-8c9d-52d594d1802b.png">

기존의 네트워크들은 receptive field size가 매우 커서 CAM과 같은 방식으로 분석한다고 해도 latent space가 pixel space와 연관짓는데 어려움이 있음
(latent representations are extracted from the whole image and unclear how the heatmaps in the latent space are related to the pixel space)

해당 논문은 이미지를 q x q patch를 stack하여 stacked patches를 이용하여 classification을 하는 bag-of-feature model인 BagNet 모델을 제시함

Patch의 logit값을 통해 설명력을 보장할 수 있어 interpretability가 좋으며, 적당한 성능을 보여줌

실제 코드를 보면 이미지를 q x q의 patch로 나누지 않고 기존의 모델들과 똑같이 256으로 resize한 후, 224 x 224로 crop하여 해당 모델 학습 및 inference함

### 2.1. Accuracy & Runtime of BagNets on ImageNet

Receptive field가 17 x 17 인 BagNet의 성능은 AlexNet과 비슷 (80.5% top-5 performance)

Receptive field가 33 x 33 인 BagNet의 성능은 top-5 performance 87.6%

ResNet-50 runtime은 570 images/s인데 반해 BagNet-q (q=9, 17, 33) runtime은 155 images/s임
(ResNet-50에 비해 downsampling 수가 적어서 더 많이 듬)

#### 2.2.1. Heatmap
<img width="803" alt="image" src="https://user-images.githubusercontent.com/49643709/167395293-a8cfd992-3ab3-4d6f-8b80-a42e8c10a4ff.png">

해당 heatmap은 다음과 같은 과정을 통해 만들어짐

1. Image를 q x q patch로 나눔 (나눈 patch에 대해 padding 해줌)
2. Patch들을 BagNet-q 모델을 통해 inference하여 정답 label에 해당하는 logit 값을 뽑음
3. Logit 값들을 224 x 224 형태로 resize
4. 원래 이미지를 canny를 통해 edge를 추출 (이미지에서의 경계선 부분들, 선택이며 필수 아님)
5. top 1%에 해당하는 logit값만 살리고 나머지는 버림
6. 이를 image edge위에 시각화

Heatmap을 보면 object의 shape 근처에 많이 분포해있으나 동물들의 경우 눈이나 다리에 heatmap이 몰려있음

#### 2.2.2. Patch with most class evidence
<img width="809" alt="image" src="https://user-images.githubusercontent.com/49643709/167395547-fc9dd0a2-6f8c-4ef7-bcdf-aed9558a3b29.png">

Logit 값이 가장 큰 image patch를 visualize

#### 2.2.3. Misclassified case
<img width="360" alt="image" src="https://user-images.githubusercontent.com/49643709/167395612-e7d74ab5-e341-4ade-80ef-95494ecc2129.png">

BagNet이 misclassified한 image에 대하여 분석함

Misclassified label을 기준으로 heatmap을 생성한 후, 해당 heatmap을 이용하여 patch를 찾음

### 2.3. Comparing the Decision-making of BagNets and High-performance DNNs

BagNet의 decision-making이 VGG-16, ResNet-50, ResNet-152, DenseNet-169와 같은 high-performance DNN과 얼마나 유사한지 실험

모델들의 logits를 이용하여 pearson correlation을 계산할 수 있겠지만 이는 decision process에 대한 정보로 보기는 힘듬

논문에서는 실험을 통해 decision process를 대략적으로 유추하며, BagNet이 생각하는 중요한 정보가 high-performance DNN에서도 중요함을 보여줌

#### 2.3.1. Image scrambling
<img width="811" alt="image" src="https://user-images.githubusercontent.com/49643709/167399771-81caafed-903a-4b70-83b0-3e285eb0598e.png">

Hidden layer activation의 gram features를 이용하여 texture synthesis algorithm으로 생성된 이미지에 대해 사람은 이를 구분해내기 어렵지만 VGG-16는 잘함
(VGG-16 is little affected, 90.1% on the clean versus 79.4% on texturised image)

이를 통해 VGG는 global shape integration을 통해 perceptual discrimination을 하는것이 아닌 local image feature의 histogram의 statistical regularities를 통해 하는 것을 알 수 있음

독특한 점은 해당 texture synthesis algorithm은 ResNet, DenseNet architecture에서 통하지 않음

#### 2.3.2. Spatially distinct image manipulations do not interact

BoF 모델은 spatial arrangement에 invariance하다는 특징을 가지고 있고, image part의 absence나 presence가 똑같은 영향을 줌

즉 이미지에 바퀴가 4개가 있다면 그 의미는 바퀴가 4개 있다는 의미로 중요도가 똑같다는 의미임

이를 식으로 표현하면 다음과 같으며, 해당 특성을 이용하여 실험을 진행

<img width="816" alt="image" src="https://user-images.githubusercontent.com/49643709/167401250-3010c249-460f-4064-b39e-ea4275ee67bf.png">
<img width="832" alt="image" src="https://user-images.githubusercontent.com/49643709/167401600-8b5b668e-3353-4c37-a969-c3cabf652483.png">

식의 좌변 (LHS)과 우변 (RHS) 사이의 pearson correlation을 측정한 실험 결과를 분석하면 다음과 같음
- Mask된 패치의 거리가 어느 수준 이상으로 넘어가면 pearson correlation이 1에 가까워짐
- 1에 가깝다는 의미는 식의 LHS와 RHS가 양의 상관관계를 가지며, 해당 식이 잘 작동한다는 의미임
- 이는 곧 BoF 모델의 특성을 가진다는 것을 말하며, 특정 pixel 이상 넘어가면 spatial한 interaction이 거의 없다는 의미임

#### 2.3.3. Error distribution
<img width="811" alt="image" src="https://user-images.githubusercontent.com/49643709/167404624-941a3e30-c281-4948-9098-aefbf9ccfab9.png">

1000가지 class에 대한 모델별 acc를 scatter plot을 통해 보여줌

Error distribution이 모델에 상관없이 충분히 일관적임
(error distribution is fairly consistent between models)

#### 2.3.4. Spatial sensitivity
<img width="825" alt="image" src="https://user-images.githubusercontent.com/49643709/167405144-ed0897b8-bad2-40a4-8ea9-262841ccfef5.png">
<img width="811" alt="image" src="https://user-images.githubusercontent.com/49643709/167405270-1790a745-f1f8-4426-8b2c-a62546f0adb4.png">

기존에 존재하는 DNN의 prediction이 어떻게 작동하는지 보여주는 attribution techniques들과 비교

가장 predictive image parts를 8 x 8 patch로 masking하며 성능이 얼마나 떨어지는지 비교

BagNet을 통해서 얻은 heatmap으로 masking을 하면 가장 성능 저하가 많이 일어났음

이 말이 의미하는 바는 BagNet이 중요하다고 생각한 image parts가 normal DNN의 classification에서도 중요함을 보여줌
(image parts that are relevant to BagNets are similarly relevant for the classification of normal DNNs)

### 2.4. Effect of Logit Thresholding
<img width="785" alt="image" src="https://user-images.githubusercontent.com/49643709/167406471-099e1e47-20d6-4d81-93bd-2d7b23ca9c3a.png">

BagNet이 각 patch의 logit의 exact value에 얼마나 sensitive한지를 2가지 방식으로 실험

첫번째 실험은 특정 threshold보다 낮으면 0, 높으면 값을 유지

두번째 실험은 특정 threshold보다 낮으면 0, 높으면 1로 바꿈

원래 BagNet 성능과 비교했을 때 크게 차이가 나지 않기에, heatmap의 amplitude가 중요하지 않다는 것을 보여줌
(amplitude of the heatmaps is not decisive)

해당 실험을 하는 이유는 BagNet이 학습, inference할 때 이미지를 patch로 나누지 않음

1장의 이미지에 대하여 그저 classification을 하는 것인데, 이는 사실 BoF 모델이라고 볼 수 있는지에 대한 의문이 있음
(그저 receptive field만 줄인 모델이기에)

BoF 모델이라면 이러한 logit값의 크기가 중요하기보다는 해당 threshold를 기준으로 있는지 없는지, 있다면 그 개수가 중요함

Binarized thresholding을 한 성능이 원래 BagNet의 성능과 차이가 나지 않기에, BagNet을 BoF 모델로 볼 수 있음을 보여주는 것임
